control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp

```diff
@@ -867,83 +905,98 @@ PidLongitudinalController::Motion PidLongitudinalController::keepBrakeBeforeStop
   return output_motion;
 }
 
-autoware_auto_planning_msgs::msg::TrajectoryPoint
-PidLongitudinalController::calcInterpolatedTargetValue(
+std::pair<autoware_auto_planning_msgs::msg::TrajectoryPoint, size_t>
+PidLongitudinalController::calcInterpolatedTrajPointAndSegment(
   const autoware_auto_planning_msgs::msg::Trajectory & traj,
   const geometry_msgs::msg::Pose & pose) const
 {
   if (traj.points.size() == 1) {
-    return traj.points.at(0);
+    return std::make_pair(traj.points.at(0), 0);
   }
 
   // apply linear interpolation
   return longitudinal_utils::lerpTrajectoryPoint(
     traj.points, pose, m_ego_nearest_dist_threshold, m_ego_nearest_yaw_threshold);
 }
 
-double PidLongitudinalController::predictedVelocityInTargetPoint(
+PidLongitudinalController::StateAfterDelay PidLongitudinalController::predictedStateAfterDelay(
   const Motion current_motion, const double delay_compensation_time) const
 {
   const double current_vel = current_motion.vel;
   const double current_acc = current_motion.acc;
-
-  if (std::fabs(current_vel) < 1e-01) {  // when velocity is low, no prediction
-    return current_vel;
-  }
-
-  const double current_vel_abs = std::fabs(current_vel);
-  if (m_ctrl_cmd_vec.size() == 0) {
-    const double pred_vel = current_vel + current_acc * delay_compensation_time;
+  double running_distance = 0.0;
+  double pred_vel = current_vel;
+  double pred_acc = current_acc;
+
+  if (m_ctrl_cmd_vec.empty() || m_current_operation_mode.mode != OperationModeState::AUTONOMOUS) {
+    // check time to stop
+    const double time_to_stop = -current_vel / current_acc;
+    const double delay_time_calculation =
+      time_to_stop > 0.0 && time_to_stop < delay_compensation_time ? time_to_stop
+                                                                   : delay_compensation_time;
+    // simple linear prediction
+    pred_vel = current_vel + current_acc * delay_time_calculation;
+    running_distance = abs(
+      delay_time_calculation * current_vel +
+      0.5 * current_acc * delay_time_calculation * delay_time_calculation);
     // avoid to change sign of current_vel and pred_vel
-    return pred_vel > 0 ? std::copysign(pred_vel, current_vel) : 0.0;
+    return StateAfterDelay{pred_vel, pred_acc, running_distance};
   }
 
-  double pred_vel = current_vel_abs;
-
-  const auto past_delay_time =
-    clock_->now() - rclcpp::Duration::from_seconds(delay_compensation_time);
   for (std::size_t i = 0; i < m_ctrl_cmd_vec.size(); ++i) {
     if ((clock_->now() - m_ctrl_cmd_vec.at(i).stamp).seconds() < m_delay_compensation_time) {
```

Comment:
Is it ok to use `m_delay_compensation_time`? Should it be `delay_compensation_time`? There are same implementations in this function.